File: pkg/interfaces/etl.go
package interfaces

type DBStorage struct {
	Name       string            `yaml:"name"`
	Type       string            `yaml:"type"`
	Connection map[string]string `yaml:"connection"`
	Query      string            `yaml:"query"`
}

type Transformation struct {
	Source string         `yaml:"source"`
	Target string         `yaml:"target"`
	Fields []FieldMapping `yaml:"fields"`
}

type FieldMapping struct {
	Source    string `yaml:"source"`
	Target    string `yaml:"target"`
	Transform string `yaml:"transform,omitempty"`
}

type ETL struct {
	Name         string      `yaml:"name"`
	Sources      []DBStorage `yaml:"sources"`
	Destinations []DBStorage `yaml:"destinations"`
}

type Pipeline struct {
	Name string `yaml:"name"`
	Jobs []ETL  `yaml:"jobs"`
}

File: web/ui/src/types/etl.ts
export interface StorageConfig {
  name: string;
  type: string;
  connection: Record<string, string>;
  connection_ref?: string;
}

export interface FieldMapping {
  source: string;
  target: string;
  transform?: string;
}

export interface TrackingSpec {
  field: string;
  storageType: string;
}

export interface PipelineConfig {
  name: string;
  sources: string[];
  target?: string; // legacy, read-only for backwards compat
  targets: string[];
  fields: FieldMapping[];
  query?: string;
  functions?: string[];
  uniqueFields?: string[];
  keepLast?: boolean;
  tracking?: TrackingSpec;
}

export interface FunctionConfig {
  name: string;
  code?: string;
}

export interface QueryConfig {
  name: string;
  sql: string;
}

export interface DeployConfig {
  schedule?: string;
  namespace?: string;
  image?: string;
}

export interface ETLConfig {
  name: string;
  sources: StorageConfig[];
  targets: StorageConfig[];
  pipelines: PipelineConfig[];
  queries: QueryConfig[];
  functions?: FunctionConfig[];
  deploy?: DeployConfig;
}
